util/profile/profile.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package profile
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"runtime/pprof"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/pprof/graph"
+	"github.com/google/pprof/measurement"
+	"github.com/google/pprof/profile"
+	"github.com/google/pprof/report"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/texttree"
+)
+
+// CPUProfileInterval represents the duration of sampling CPU
+var CPUProfileInterval = 30 * time.Second
+
+// Collector is used to collect the profile results
+type Collector struct {
+	Rows [][]types.Datum
+}
+
+type perfNode struct {
+	Name      string
+	Location  string
+	Cum       int64
+	CumFormat string
+	Percent   string
+	Children  []*perfNode
+}
+
+func (c *Collector) collect(node *perfNode, depth int64, indent string, rootChild int, parentCur int64, isLastChild bool) {
+	row := types.MakeDatums(
+		texttree.PrettyIdentifier(node.Name, indent, isLastChild),
+		node.Percent,
+		strings.TrimSpace(measurement.Percentage(node.Cum, parentCur)),
+		rootChild,
+		depth,
+		node.Location,
+	)
+	c.Rows = append(c.Rows, row)
+
+	indent4Child := texttree.Indent4Child(indent, isLastChild)
+	for i, child := range node.Children {
```

Comment:
Is the report graph guaranteed to be acyclic? Maybe it's better to add some protections in case of cycle happened(say, some bugs occurred in Go profiling)

Reply:
Yes, it's acyclic.